File: src/temporal/activities/deployments.py
"""Activities for Temporal Worker Deployment and Docker cleanup."""

import http.client
import json
import os
import re
import socket
from typing import Any
from urllib.parse import quote, urlencode
from urllib.request import urlopen

from temporalio import activity
from temporalio.api.deployment.v1 import WorkerDeploymentInfo
from temporalio.api.enums.v1 import (
    VersionDrainageStatus,
    WorkerDeploymentVersionStatus,
)
from temporalio.api.workflowservice.v1 import (
    DescribeWorkerDeploymentRequest,
    ListWorkflowExecutionsRequest,
    SetWorkerDeploymentCurrentVersionRequest,
)
from temporalio.client import Client

from bot.config import TASK_QUEUE, TEMPORAL_HOST, TEMPORAL_NAMESPACE
from temporal.deployment_models import (
    DeploymentHealthCheck,
    DeploymentRollbackResult,
    DockerCleanupResult,
    DockerContainerState,
    SdkMetricsSnapshot,
    TemporalExecutionSummary,
    WorkerDeploymentState,
    WorkerDeploymentVersionState,
)

DEPLOYMENT_LABEL = "com.reddit-bots.deployment-name"
BUILD_ID_LABEL = "com.reddit-bots.build-id"
IMAGE_LABEL = "com.reddit-bots.image"
DEFAULT_DOCKER_SOCKET = "/var/run/docker.sock"
DEPLOYMENT_ACTIVITY_TYPES = {
    "collect_sdk_metrics",
    "count_completed_workflows_for_deployment",
    "count_failed_workflows_for_deployment",
    "describe_worker_deployment",
    "list_deployment_containers",
    "remove_deployment_container",
    "rollback_worker_deployment",
}
PROMETHEUS_SAMPLE_RE = re.compile(
    r"^(?P<name>[a-zA-Z_:][a-zA-Z0-9_:]*)(?:\{(?P<labels>[^}]*)\})?\s+"
    r"(?P<value>[-+0-9.eE]+)"
)


class UnixSocketHTTPConnection(http.client.HTTPConnection):
    """Minimal HTTP client for the Docker Engine Unix socket API."""

    def __init__(self, socket_path: str):
        super().__init__("localhost")
        self.socket_path = socket_path

    def connect(self) -> None:
        sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        sock.connect(self.socket_path)
        self.sock = sock


def _docker_socket_path() -> str:
    return os.getenv("DOCKER_SOCKET", DEFAULT_DOCKER_SOCKET)


def _docker_request(
    method: str,
    path: str,
    *,
    body: dict[str, Any] | None = None,
    ok_statuses: set[int] | None = None,
) -> Any:
    socket_path = _docker_socket_path()
    if not os.path.exists(socket_path):
        raise RuntimeError(f"Docker socket not found: {socket_path}")

    ok_statuses = ok_statuses or {200, 201, 204}
    encoded_body = None
    headers = {}
    if body is not None:
        encoded_body = json.dumps(body).encode("utf-8")
        headers["Content-Type"] = "application/json"

    conn = UnixSocketHTTPConnection(socket_path)
    try:
        conn.request(method, path, body=encoded_body, headers=headers)
        response = conn.getresponse()
        response_body = response.read()
    finally:
        conn.close()

    if response.status not in ok_statuses:
        detail = response_body.decode("utf-8", errors="replace")
        raise RuntimeError(
            f"Docker API {method} {path} failed with {response.status}: {detail}"
        )

    if not response_body:
        return None
    return json.loads(response_body.decode("utf-8"))


def _enum_suffix(enum_wrapper: Any, value: int, prefix: str) -> str:
    name = enum_wrapper.Name(value)
    if name.startswith(prefix):
        return name.removeprefix(prefix)
    return name


def _visibility_query_string(value: str) -> str:
    """Escape a string for use in a Temporal visibility query."""
    return value.replace("\\", "\\\\").replace('"', '\\"')


def _version_summary_to_state(
    summary: WorkerDeploymentInfo.WorkerDeploymentVersionSummary,
) -> WorkerDeploymentVersionState:
    build_id = summary.deployment_version.build_id
    if not build_id:
        build_id = summary.version.rsplit(".", 1)[-1]

    drainage_status = None
    if summary.drainage_status:
        drainage_status = _enum_suffix(
            VersionDrainageStatus,
            summary.drainage_status,
            "VERSION_DRAINAGE_STATUS_",
        )
    elif summary.HasField("drainage_info") and summary.drainage_info.status:
        drainage_status = _enum_suffix(
            VersionDrainageStatus,
            summary.drainage_info.status,
            "VERSION_DRAINAGE_STATUS_",
        )

    return WorkerDeploymentVersionState(
        build_id=build_id,
        status=_enum_suffix(
            WorkerDeploymentVersionStatus,
            summary.status,
            "WORKER_DEPLOYMENT_VERSION_STATUS_",
        ),
        drainage_status=drainage_status,
    )


def _labels_from_prometheus_sample(raw_labels: str | None) -> dict[str, str]:
    if not raw_labels:
        return {}

    labels: dict[str, str] = {}
    label_pattern = r'([a-zA-Z_][a-zA-Z0-9_]*)="((?:\\.|[^"])*)"'
    for item in re.finditer(label_pattern, raw_labels):
        labels[item.group(1)] = item.group(2).replace(r"\"", '"')
    return labels


def _sample_matches(
    labels: dict[str, str],
    *,
    namespace: str,
    task_queue: str,
    subreddit_name: str,
    exclude_activity_types: set[str] | None = None,
) -> bool:
    if (
        exclude_activity_types
        and labels.get("activity_type") in exclude_activity_types
    ):
        return False

    expected = {
        "namespace": namespace,
        "task_queue": task_queue,
        "subreddit": subreddit_name,
    }
    for key, value in expected.items():
        if key in labels and labels[key] != value:
            return False
    return True


def _sum_prometheus_counter(
    text: str,
    metric_names: set[str],
    *,
    namespace: str,
    task_queue: str,
    subreddit_name: str,
    exclude_activity_types: set[str] | None = None,
) -> float:
    total = 0.0
    for line in text.splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue

        match = PROMETHEUS_SAMPLE_RE.match(line)
        if match is None:
            continue

        name = match.group("name")
        if name not in metric_names and not (
            name.endswith("_total") and name.removesuffix("_total") in metric_names
        ):
            continue

        labels = _labels_from_prometheus_sample(match.group("labels"))
        if not _sample_matches(
            labels,
            namespace=namespace,
            task_queue=task_queue,
            subreddit_name=subreddit_name,
            exclude_activity_types=exclude_activity_types,
        ):
            continue

        total += float(match.group("value"))
    return total


@activity.defn
async def describe_worker_deployment(deployment_name: str) -> WorkerDeploymentState:
    """Describe a Temporal Worker Deployment and its versions."""
    client = await Client.connect(TEMPORAL_HOST, namespace=TEMPORAL_NAMESPACE)
    response = await client.workflow_service.describe_worker_deployment(
        DescribeWorkerDeploymentRequest(
            namespace=TEMPORAL_NAMESPACE,
            deployment_name=deployment_name,
        ),
        retry=True,
    )

    info = response.worker_deployment_info
    return WorkerDeploymentState(
        deployment_name=deployment_name,
        versions=[
            _version_summary_to_state(summary)
            for summary in info.version_summaries
        ],
    )


async def _count_workflows_for_deployment(
    deployment_name: str,
    build_id: str,
    since_time_iso: str,
    status: str,
) -> TemporalExecutionSummary:
    """Count recent workflows attributed to a deployment build by status."""
    client = await Client.connect(TEMPORAL_HOST, namespace=TEMPORAL_NAMESPACE)
    deployment_version = f"{deployment_name}:{build_id}"
    query = (
        f'ExecutionStatus = "{status}" '
        f'AND CloseTime >= "{since_time_iso}" '
        f'AND TemporalWorkerDeployment = "{_visibility_query_string(deployment_name)}" '
        f'AND TemporalWorkerDeploymentVersion = '
        f'"{_visibility_query_string(deployment_version)}"'
    )
    next_page_token = b""
    workflow_ids: list[str] = []

    while True:
        response = await client.workflow_service.list_workflow_executions(
            ListWorkflowExecutionsRequest(
                namespace=TEMPORAL_NAMESPACE,
                page_size=100,
                next_page_token=next_page_token,
                query=query,
            ),
            retry=True,
        )

        workflow_ids.extend(
            execution_info.execution.workflow_id
            for execution_info in response.executions
        )

        next_page_token = response.next_page_token
        if not next_page_token:
            break

    return TemporalExecutionSummary(
        count=len(workflow_ids),
        workflow_ids=workflow_ids[:20],
    )


@activity.defn
async def count_failed_workflows_for_deployment(
    deployment_name: str,
    build_id: str,
    since_time_iso: str,
) -> TemporalExecutionSummary:
    """Count recently failed workflows attributed to a deployment build."""
    return await _count_workflows_for_deployment(
        deployment_name,
        build_id,
        since_time_iso,
        "Failed",
    )


@activity.defn
async def count_completed_workflows_for_deployment(
    deployment_name: str,
    build_id: str,
    since_time_iso: str,
) -> TemporalExecutionSummary:
    """Count recently completed workflows attributed to a deployment build."""
    return await _count_workflows_for_deployment(
        deployment_name,
        build_id,
        since_time_iso,
        "Completed",
    )


@activity.defn
async def collect_sdk_metrics(
    health_check: DeploymentHealthCheck,
    subreddit_name: str,
) -> SdkMetricsSnapshot:
    """Collect Temporal SDK failure counters from the worker metrics endpoint."""
    metrics_url = health_check.metrics_url or os.getenv("DEPLOYMENT_HEALTH_METRICS_URL")
    if not metrics_url:
        bind_address = os.getenv("TEMPORAL_SDK_METRICS_BIND_ADDRESS", "").strip()
        if bind_address:
            port = bind_address.rsplit(":", 1)[-1]
            metrics_url = f"http://127.0.0.1:{port}/metrics"

    if not metrics_url:
        raise RuntimeError("No SDK metrics URL configured")

    with urlopen(metrics_url, timeout=5) as response:
        metrics_text = response.read().decode("utf-8", errors="replace")

    common = {
        "namespace": TEMPORAL_NAMESPACE,
        "task_queue": TASK_QUEUE,
        "subreddit_name": subreddit_name,
    }
    return SdkMetricsSnapshot(
        workflow_completed=_sum_prometheus_counter(
            metrics_text,
            {"temporal_workflow_completed"},
            **common,
        ),
        activity_completed=_sum_prometheus_counter(
            metrics_text,
            {
                "temporal_activity_execution_completed",
                "temporal_activity_execution_latency_count",
            },
            exclude_activity_types=DEPLOYMENT_ACTIVITY_TYPES,
            **common,
        ),
        workflow_failed=_sum_prometheus_counter(
            metrics_text,
            {"temporal_workflow_failed"},
            **common,
        ),
        activity_failed=_sum_prometheus_counter(
            metrics_text,
            {"temporal_activity_execution_failed"},
            **common,
        ),
        workflow_task_failed=_sum_prometheus_counter(
            metrics_text,
            {
                "temporal_workflow_task_execution_failed",
                "temporal_workflow_task_failed",
            },
            **common,
        ),
    )


@activity.defn
async def list_deployment_containers(
    deployment_name: str,
) -> list[DockerContainerState]:
    """List Docker containers associated with a Temporal deployment."""
    filters = json.dumps({"label": [f"{DEPLOYMENT_LABEL}={deployment_name}"]})
    query = urlencode({"all": "true", "filters": filters})
    containers = _docker_request("GET", f"/containers/json?{query}") or []

    result: list[DockerContainerState] = []
    for container in containers:
        labels = container.get("Labels") or {}
        build_id = labels.get(BUILD_ID_LABEL)
        if not build_id:
            continue

        names = container.get("Names") or []
        name = names[0].lstrip("/") if names else container["Id"][:12]
        result.append(
            DockerContainerState(
                id=container["Id"],
                name=name,
                build_id=build_id,
                image=labels.get(IMAGE_LABEL) or container.get("Image"),
                state=container.get("State"),
                status=container.get("Status"),
            )
        )

    return result


@activity.defn
async def rollback_worker_deployment(
    deployment_name: str,
    rollback_build_id: str,
    failed_container: DockerContainerState | None = None,
) -> DeploymentRollbackResult:
    """Set a Worker Deployment back to a previous build and remove failed container."""
    client = await Client.connect(TEMPORAL_HOST, namespace=TEMPORAL_NAMESPACE)
    await client.workflow_service.set_worker_deployment_current_version(
        SetWorkerDeploymentCurrentVersionRequest(
            namespace=TEMPORAL_NAMESPACE,
            deployment_name=deployment_name,
            build_id=rollback_build_id,
            identity="reddit-bots-deployment-health",
        ),
        retry=True,
    )

    result = DeploymentRollbackResult(
        deployment_name=deployment_name,
        rollback_build_id=rollback_build_id,
    )
    if failed_container is not None:
        container_ref = quote(failed_container.id, safe="")
        _docker_request(
            "DELETE",
            f"/containers/{container_ref}?force=true&v=true",
            ok_statuses={204, 404},
        )
        result.container_removed = True
    return result


@activity.defn
async def remove_deployment_container(
    container: DockerContainerState,
) -> DockerCleanupResult:
    """Remove a drained deployment container and its now-unused image."""
    result = DockerCleanupResult(container_name=container.name)
    container_ref = quote(container.id, safe="")

    _docker_request(
        "DELETE",
        f"/containers/{container_ref}?force=true&v=true",
        ok_statuses={204, 404},
    )
    result.container_removed = True

    prune_filters = json.dumps({"label": [f"{BUILD_ID_LABEL}={container.build_id}"]})
    prune_query = urlencode({"filters": prune_filters})
    prune_response = _docker_request("POST", f"/containers/prune?{prune_query}") or {}
    deleted = prune_response.get("ContainersDeleted") or []
    result.pruned_containers = len(deleted)

    if container.image:
        image_ref = quote(container.image, safe="")
        try:
            _docker_request(
                "DELETE",
                f"/images/{image_ref}?force=false&noprune=false",
                ok_statuses={200, 202, 204, 404},
            )
            result.image_removed = True
        except RuntimeError as err:
            # A 409 means another container still uses the image. Report it but
            # do not fail the workflow; a future deployment cleanup can retry.
            result.image_remove_error = str(err)

    return result

File: internal/activities/deployments.go
package activities

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"strings"
	"time"

	"go.temporal.io/api/workflowservice/v1"
	"go.temporal.io/sdk/client"

	"github.com/mikeacjones/reddit-trade-confirmation-bot/internal/deployment"
)

const (
	deploymentLabel     = "com.reddit-bots.deployment-name"
	buildIDLabel        = "com.reddit-bots.build-id"
	imageLabel          = "com.reddit-bots.image"
	defaultDockerSocket = "/var/run/docker.sock"
)

var deploymentActivityTypes = map[string]struct{}{
	"collect_sdk_metrics":                        {},
	"count_completed_workflows_for_deployment":   {},
	"count_failed_workflows_for_deployment":      {},
	"describe_worker_deployment":                 {},
	"list_deployment_containers":                 {},
	"remove_deployment_container":                {},
	"rollback_worker_deployment":                 {},
}

var prometheusSampleRE = regexp.MustCompile(
	`^(?P<name>[a-zA-Z_:][a-zA-Z0-9_:]*)(?:\{(?P<labels>[^}]*)\})?\s+(?P<value>[-+0-9.eE]+)`,
)

type unixDialer struct{ path string }

func (d unixDialer) Dial(_, _ string) (net.Conn, error) {
	return net.Dial("unix", d.path)
}

func dockerSocketPath() string {
	if p := os.Getenv("DOCKER_SOCKET"); p != "" {
		return p
	}
	return defaultDockerSocket
}

func dockerRequest(method, path string, body any, okStatuses map[int]struct{}) (json.RawMessage, error) {
	socketPath := dockerSocketPath()
	if _, err := os.Stat(socketPath); err != nil {
		return nil, fmt.Errorf("Docker socket not found: %s", socketPath)
	}
	if okStatuses == nil {
		okStatuses = map[int]struct{}{200: {}, 201: {}, 204: {}}
	}
	var bodyReader io.Reader
	headers := http.Header{}
	if body != nil {
		b, err := json.Marshal(body)
		if err != nil {
			return nil, err
		}
		bodyReader = strings.NewReader(string(b))
		headers.Set("Content-Type", "application/json")
	}
	transport := &http.Transport{Dial: unixDialer{path: socketPath}.Dial}
	httpClient := &http.Client{Transport: transport, Timeout: 30 * time.Second}
	req, err := http.NewRequest(method, "http://localhost"+path, bodyReader)
	if err != nil {
		return nil, err
	}
	req.Header = headers
	resp, err := httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	raw, _ := io.ReadAll(resp.Body)
	if _, ok := okStatuses[resp.StatusCode]; !ok {
		return nil, fmt.Errorf("Docker API %s %s failed with %d: %s", method, path, resp.StatusCode, raw)
	}
	if len(raw) == 0 {
		return nil, nil
	}
	return json.RawMessage(raw), nil
}

func visibilityEscape(value string) string {
	return strings.ReplaceAll(strings.ReplaceAll(value, `\`, `\\`), `"`, `\"`)
}

// DescribeWorkerDeployment describes a Temporal Worker Deployment and its versions.
func (a *Activities) DescribeWorkerDeployment(ctx context.Context, deploymentName string) (deployment.WorkerDeploymentState, error) {
	handle := a.Temporal.WorkerDeploymentClient().GetHandle(deploymentName)
	resp, err := handle.Describe(ctx, client.WorkerDeploymentDescribeOptions{})
	if err != nil {
		return deployment.WorkerDeploymentState{}, err
	}
	currentBuild := ""
	if resp.Info.RoutingConfig.CurrentVersion != nil {
		currentBuild = resp.Info.RoutingConfig.CurrentVersion.BuildID
	}
	rampingBuild := ""
	if resp.Info.RoutingConfig.RampingVersion != nil {
		rampingBuild = resp.Info.RoutingConfig.RampingVersion.BuildID
	}
	versions := make([]deployment.WorkerVersionState, 0, len(resp.Info.VersionSummaries))
	for _, summary := range resp.Info.VersionSummaries {
		buildID := summary.Version.BuildID
		var drainage *string
		if summary.DrainageStatus != client.WorkerDeploymentVersionDrainageStatusUnspecified {
			s := drainageStatusString(summary.DrainageStatus)
			drainage = &s
		}
		versions = append(versions, deployment.WorkerVersionState{
			BuildID:        buildID,
			Status:         deriveVersionStatus(buildID, currentBuild, rampingBuild, summary.DrainageStatus),
			DrainageStatus: drainage,
		})
	}
	return deployment.WorkerDeploymentState{
		DeploymentName: deploymentName,
		Versions:       versions,
	}, nil
}

func deriveVersionStatus(buildID, currentBuild, rampingBuild string, drainage client.WorkerDeploymentVersionDrainageStatus) string {
	if buildID != "" && buildID == currentBuild {
		return "CURRENT"
	}
	if buildID != "" && buildID == rampingBuild {
		return "RAMPING"
	}
	switch drainage {
	case client.WorkerDeploymentVersionDrainageStatusDraining:
		return "DRAINING"
	case client.WorkerDeploymentVersionDrainageStatusDrained:
		return "DRAINED"
	default:
		return "INACTIVE"
	}
}

func drainageStatusString(s client.WorkerDeploymentVersionDrainageStatus) string {
	switch s {
	case client.WorkerDeploymentVersionDrainageStatusDraining:
		return "DRAINING"
	case client.WorkerDeploymentVersionDrainageStatusDrained:
		return "DRAINED"
	default:
		return "UNSPECIFIED"
	}
}

func (a *Activities) countWorkflowsForDeployment(
	ctx context.Context,
	deploymentName, buildID, sinceTimeISO, status string,
) (deployment.TemporalExecutionSummary, error) {
	deploymentVersion := deploymentName + ":" + buildID
	query := fmt.Sprintf(
		`ExecutionStatus = "%s" AND CloseTime >= "%s" AND TemporalWorkerDeployment = "%s" AND TemporalWorkerDeploymentVersion = "%s"`,
		status, sinceTimeISO, visibilityEscape(deploymentName), visibilityEscape(deploymentVersion),
	)
	var workflowIDs []string
	var nextPageToken []byte
	for {
		resp, err := a.Temporal.WorkflowService().ListWorkflowExecutions(ctx, &workflowservice.ListWorkflowExecutionsRequest{
			Namespace:     a.Cfg.TemporalNamespace,
			PageSize:      100,
			NextPageToken: nextPageToken,
			Query:         query,
		})
		if err != nil {
			return deployment.TemporalExecutionSummary{}, err
		}
		for _, exec := range resp.Executions {
			workflowIDs = append(workflowIDs, exec.Execution.WorkflowId)
		}
		nextPageToken = resp.NextPageToken
		if len(nextPageToken) == 0 {
			break
		}
	}
	ids := workflowIDs
	if len(ids) > 20 {
		ids = ids[:20]
	}
	return deployment.TemporalExecutionSummary{Count: len(workflowIDs), WorkflowIDs: ids}, nil
}

// CountFailedWorkflowsForDeployment counts recently failed workflows for a build.
func (a *Activities) CountFailedWorkflowsForDeployment(ctx context.Context, deploymentName, buildID, sinceTimeISO string) (deployment.TemporalExecutionSummary, error) {
	return a.countWorkflowsForDeployment(ctx, deploymentName, buildID, sinceTimeISO, "Failed")
}

// CountCompletedWorkflowsForDeployment counts recently completed workflows for a build.
func (a *Activities) CountCompletedWorkflowsForDeployment(ctx context.Context, deploymentName, buildID, sinceTimeISO string) (deployment.TemporalExecutionSummary, error) {
	return a.countWorkflowsForDeployment(ctx, deploymentName, buildID, sinceTimeISO, "Completed")
}

// CollectSDKMetrics collects Temporal SDK failure counters from the metrics endpoint.
func (a *Activities) CollectSDKMetrics(ctx context.Context, healthCheck deployment.HealthCheck, subredditName string) (deployment.SDKMetricsSnapshot, error) {
	metricsURL := ""
	if healthCheck.MetricsURL != nil {
		metricsURL = *healthCheck.MetricsURL
	}
	if metricsURL == "" {
		metricsURL = os.Getenv("DEPLOYMENT_HEALTH_METRICS_URL")
	}
	if metricsURL == "" {
		bind := strings.TrimSpace(os.Getenv("TEMPORAL_SDK_METRICS_BIND_ADDRESS"))
		if bind != "" {
			parts := strings.Split(bind, ":")
			port := parts[len(parts)-1]
			metricsURL = "http://127.0.0.1:" + port + "/metrics"
		}
	}
	if metricsURL == "" {
		return deployment.SDKMetricsSnapshot{}, fmt.Errorf("No SDK metrics URL configured")
	}
	httpClient := &http.Client{Timeout: 5 * time.Second}
	resp, err := httpClient.Get(metricsURL)
	if err != nil {
		return deployment.SDKMetricsSnapshot{}, err
	}
	defer resp.Body.Close()
	raw, err := io.ReadAll(resp.Body)
	if err != nil {
		return deployment.SDKMetricsSnapshot{}, err
	}
	text := string(raw)
	common := metricFilter{
		namespace:     a.Cfg.TemporalNamespace,
		taskQueue:     a.Cfg.TaskQueue,
		subredditName: subredditName,
	}
	return deployment.SDKMetricsSnapshot{
		WorkflowCompleted: sumPrometheus(text, []string{"temporal_workflow_completed"}, common, nil),
		ActivityCompleted: sumPrometheus(text, []string{
			"temporal_activity_execution_completed",
			"temporal_activity_execution_latency_count",
		}, common, deploymentActivityTypes),
		WorkflowFailed: sumPrometheus(text, []string{"temporal_workflow_failed"}, common, nil),
		ActivityFailed: sumPrometheus(text, []string{"temporal_activity_execution_failed"}, common, nil),
		WorkflowTaskFailed: sumPrometheus(text, []string{
			"temporal_workflow_task_execution_failed",
			"temporal_workflow_task_failed",
		}, common, nil),
	}, nil
}

type metricFilter struct {
	namespace, taskQueue, subredditName string
}

func sumPrometheus(text string, names []string, f metricFilter, exclude map[string]struct{}) float64 {
	nameSet := map[string]struct{}{}
	for _, n := range names {
		nameSet[n] = struct{}{}
	}
	var total float64
	for _, line := range strings.Split(text, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		m := prometheusSampleRE.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		name := m[1]
		base := strings.TrimSuffix(name, "_total")
		if _, ok := nameSet[name]; !ok {
			if _, ok2 := nameSet[base]; !ok2 {
				continue
			}
		}
		labels := parsePromLabels(m[2])
		if exclude != nil {
			if at, ok := labels["activity_type"]; ok {
				if _, skip := exclude[at]; skip {
					continue
				}
			}
		}
		if v, ok := labels["namespace"]; ok && v != f.namespace {
			continue
		}
		if v, ok := labels["task_queue"]; ok && v != f.taskQueue {
			continue
		}
		if v, ok := labels["subreddit"]; ok && v != f.subredditName {
			continue
		}
		var val float64
		fmt.Sscanf(m[3], "%f", &val)
		total += val
	}
	return total
}

func parsePromLabels(raw string) map[string]string {
	out := map[string]string{}
	if raw == "" {
		return out
	}
	re := regexp.MustCompile(`([a-zA-Z_][a-zA-Z0-9_]*)="((?:\\.|[^"])*)"`)
	for _, m := range re.FindAllStringSubmatch(raw, -1) {
		out[m[1]] = strings.ReplaceAll(m[2], `\"`, `"`)
	}
	return out
}

// ListDeploymentContainers lists Docker containers for a deployment.
func (a *Activities) ListDeploymentContainers(ctx context.Context, deploymentName string) ([]deployment.DockerContainerState, error) {
	filters, _ := json.Marshal(map[string][]string{"label": {deploymentLabel + "=" + deploymentName}})
	q := url.Values{"all": {"true"}, "filters": {string(filters)}}.Encode()
	raw, err := dockerRequest("GET", "/containers/json?"+q, nil, nil)
	if err != nil {
		return nil, err
	}
	var containers []struct {
		ID     string            `json:"Id"`
		Names  []string          `json:"Names"`
		Image  string            `json:"Image"`
		State  string            `json:"State"`
		Status string            `json:"Status"`
		Labels map[string]string `json:"Labels"`
	}
	if err := json.Unmarshal(raw, &containers); err != nil {
		return nil, err
	}
	var result []deployment.DockerContainerState
	for _, c := range containers {
		buildID := c.Labels[buildIDLabel]
		if buildID == "" {
			continue
		}
		name := c.ID[:12]
		if len(c.Names) > 0 {
			name = strings.TrimPrefix(c.Names[0], "/")
		}
		img := c.Labels[imageLabel]
		if img == "" {
			img = c.Image
		}
		state, status := c.State, c.Status
		result = append(result, deployment.DockerContainerState{
			ID:      c.ID,
			Name:    name,
			BuildID: buildID,
			Image:   &img,
			State:   &state,
			Status:  &status,
		})
	}
	return result, nil
}

// RollbackWorkerDeployment sets current version back and removes the failed container.
func (a *Activities) RollbackWorkerDeployment(ctx context.Context, deploymentName, rollbackBuildID string, failedContainer *deployment.DockerContainerState) (deployment.RollbackResult, error) {
	handle := a.Temporal.WorkerDeploymentClient().GetHandle(deploymentName)
	_, err := handle.SetCurrentVersion(ctx, client.WorkerDeploymentSetCurrentVersionOptions{
		BuildID:  rollbackBuildID,
		Identity: "reddit-bots-deployment-health",
	})
	if err != nil {
		return deployment.RollbackResult{}, err
	}
	result := deployment.RollbackResult{
		DeploymentName:  deploymentName,
		RollbackBuildID: rollbackBuildID,
	}
	if failedContainer != nil {
		ref := url.PathEscape(failedContainer.ID)
		_, err := dockerRequest("DELETE", "/containers/"+ref+"?force=true&v=true", nil, map[int]struct{}{204: {}, 404: {}})
		if err != nil {
			return result, err
		}
		result.ContainerRemoved = true
	}
	return result, nil
}

// RemoveDeploymentContainer removes a drained deployment container and unused image.
func (a *Activities) RemoveDeploymentContainer(ctx context.Context, container deployment.DockerContainerState) (deployment.DockerCleanupResult, error) {
	result := deployment.DockerCleanupResult{ContainerName: container.Name}
	ref := url.PathEscape(container.ID)
	if _, err := dockerRequest("DELETE", "/containers/"+ref+"?force=true&v=true", nil, map[int]struct{}{204: {}, 404: {}}); err != nil {
		return result, err
	}
	result.ContainerRemoved = true

	filters, _ := json.Marshal(map[string][]string{"label": {buildIDLabel + "=" + container.BuildID}})
	q := url.Values{"filters": {string(filters)}}.Encode()
	raw, err := dockerRequest("POST", "/containers/prune?"+q, nil, nil)
	if err == nil && raw != nil {
		var prune struct {
			ContainersDeleted []string `json:"ContainersDeleted"`
		}
		_ = json.Unmarshal(raw, &prune)
		result.PrunedContainers = len(prune.ContainersDeleted)
	}

	if container.Image != nil && *container.Image != "" {
		imgRef := url.PathEscape(*container.Image)
		if _, err := dockerRequest("DELETE", "/images/"+imgRef+"?force=false&noprune=false", nil, map[int]struct{}{200: {}, 202: {}, 204: {}, 404: {}}); err != nil {
			msg := err.Error()
			result.ImageRemoveError = &msg
		} else {
			result.ImageRemoved = true
		}
	}
	return result, nil
}
